### PR TITLE
Implement support for ragged tensors in mean_squared_error loss function.

### PR DESCRIPTION
**Pull request by pedro-r-marques**
First cut at addressing Issue #45403.

This PR adds support for RaggedTensors to "mse" loss. Tensors with shapes of (batch, ragged_dimension) or (batch, ragged_dimension, feature dims) are handled on a per batch basis.

### CODE REVIEW

File: tensorflow/python/keras/utils/losses_utils.py [hunk 1]
```diff
@@ -255,7 +255,8 @@ def compute_weighted_loss(losses,
     # to multiple replicas. Used only for estimator + v1 optimizer flow.
     ops.get_default_graph()._last_loss_reduction = reduction  # pylint: disable=protected-access
 
-    if not isinstance(losses, keras_tensor.KerasTensor):
+    if not isinstance(
```
Comment (tomerk) [inline tensorflow/python/keras/utils/losses_utils.py:258]: This is a good example of a place that prevents Keras from supporting CompositeTensor losses in general right now (regardless of what behavior that loss should have)

Rather than hard-coding `RaggedTensor` here could you try using `composite_tensor.CompositeTensor`? That'll increase the chance that this will generalize to other extension types beyond just RaggedTensors.

File: tensorflow/python/keras/losses.py [hunk 2]
```diff
@@ -250,7 +289,13 @@ def call(self, y_true, y_pred):
     """
     if tensor_util.is_tf_type(y_pred) and tensor_util.is_tf_type(y_true):
       y_pred, y_true = losses_utils.squeeze_or_expand_dimensions(y_pred, y_true)
-    ag_fn = autograph.tf_convert(self.fn, ag_ctx.control_status_ctx())
+
+    if (isinstance(y_true, ragged_tensor.RaggedTensor) and
+        isinstance(y_pred, ragged_tensor.RaggedTensor)):
+      loss_fn = self._eval_ragged_tensors
```
Comment (tomerk) [inline tensorflow/python/keras/losses.py:295]: This and the `_eval_ragged_tensors` method above make me nervous.

It's baking in some pretty subtle decisions about how to handle RaggedTensors of specific shapes in different ways into what's meant to be a fairly generic/simple component. It also means the behavior of the losses wouldn't match the corresponding `keras.metrics.*` apis that they use under the hood.

We've gotten burned by this sort of thing in Keras a lot, and it tends to cause lingering complexity and hard-to-track subtle inconsistencies between similar apis. And, as the github issue discussion on https://github.com/tensorflow/tensorflow/issues/45403 shows there's more than one 'reasonable' sets of behaviors users might want.

I think ideally the LossFunctionWrapper would continue to be generic and not do anything special for RaggedTensors that the corresponding `keras.metrics.mean_squared_error`/etc. apis don't do. Instead, we fix anything that doesn't currently accept CompositeTensors/RaggedTensors to be able to accept them.

-------------------
I think these are all sensible options for how to do that:

1. If the implementation of the `keras.metrics.*` losses in this file are sensible for raggedtensors but it's just the conversion that causes them to break, we can choose to only call the `convert_to_tensor` calls if the input is not already a CompositeTensor.

2. If RaggedTensors need to be handled in a single consistent way and the above isn't enough, we can use the extensiontype dispatching mechanism to add implementations for ragged inputs to these Keras apis. (The same mechanism raggedtensor uses to implement how it behaves for built-in tf apis: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/ragged/ragged_dispatch.py )

3. If there's multiple "reasonable" ways for losses to handle raggedtensors and users might want to switch between them, we shouldn't bake that functionality into the built-in loss apis. Instead, we should add distinct `ragged` losses that support treating raggedtensors in different ways according to what users specify.

**Review by tomerk — COMMENTED** [review]
Thank you for looking into this!
I have some thoughts on how we might be able to make this more sustainable w/o adding extra complexity to already-existing things (that are already more complex than we would like)

File: tensorflow/python/keras/losses.py (hunk 2 above)
Comment (pedro-r-marques) [reply tensorflow/python/keras/losses.py:295]: My understanding / assumptions:
- The builtin keras api "mse" metric will perform a mean operation over the last axis; the mean over the batch is performed by the reduction function (as the the loss class reduction argument).
- The keras losses API assumes the ability to specify per batch weights;

My expectation would be that when one specifies a model with a ragged tensor output and a "mse" loss that the correct behaviour would be that the loss would calculate the mean over the batch and than the reduction function would be able to apply weights per batch. One of the tests that I added illustrates this well:
```
   y_true = ragged_factory_ops.constant([[1., 1., 9.], [2., 5.]])
    y_pred = ragged_factory_ops.constant([[4., 1., 8.], [12., 3.]])
    sample_weight = constant_op.constant([1.2, 0.5])
    loss = mse_obj(y_true, y_pred, sample_weight=sample_weight)

    # mse = [((4 - 1)^2 + (8 - 9)^2) / 3, ((12 - 2)^2 + (3 - 5)^2) / 2]
    # mse = [3.(3), 52]
    # weighted_mse = [3.(3) * 1.2, 52 * 0.5] = [4, 26]
    # reduced_weighted_mse = (4 + 26) / 2 =
    self.assertAllClose(self.evaluate(loss), 15, 1e-2)
```

The reason I bring this up is that if one assumes that this should be the default API behaviour then that leads to the requirement that the loss be computed by evaluating the mse function over the batches (e.g. via ```ragged_map_ops.map_fn```); If one where to convert the above tensors from ragged to dense the dense tensor in the example above would have a shape of (2, 3) which would lead to the mean of batch 1 having a different value. This effect would be much more pronounced when dealing with a significant variance in the ragged sequence dimension (the use case that motivates me is that I often deal with a bi-modal distribution of sequence lengths).

I believe this argues for option 2 above. i.e. it is not sufficient to just convert a CompositeTensor to a dense tensor; however there is a reasonable expectation for how to adapt the loss functions in a common way such that the loss API can still preserve it common ability to add per batch weights before reducing the the per batch loss values.

In general, my assumptions:
 - The current behaviour is undesirable; the fit method just fails with a hard to diagnose message;
 - The most common scenario is that the ragged dimension is a sequence dimension that follows the batch; there could be subsequent dimensions after or not.
 - It would be desirable to handle this consistently for all the loss functions rather than have to implement support in each... e.g. "mse", "mae", "binary_crossentrophy".
 - As with dense tensors, the API should support the ability to add per batch weights by default; there could be other loss reduction options that don't respect batch boundary but that should be the default.

I was hoping to get feedback on my assumptions above; please let me know if my conclusions are incorrect.

At the moment, I'm inclined to think that there should be generic support for ragged tensors in the loss library (different than just converting to a dense tensor).

When you outline option 2 above, did you have in mind to create a wrapper per each of the predefined loss functions ? i.e. to create a ragged_mse loss  that is then added via dispatch ? I would think that would make the code much harder to follow than the option I proposed. It is quite possible that the proposed function ```_eval_ragged_tensors``` is incomplete and does not cover all desirable use cases; but it does have the advantage that it is common across different losses and that the code path is clear.

Comment (tomerk) [reply tensorflow/python/keras/losses.py:295]: Yeah I think your assumptions are about right.

Above I wasn't suggesting converting them to a dense tensor, I was saying it may be possible to not force any conversion at all (generically accepting all compositetensors w/o conversion) and rely on the lower-level ops/losses being able to handle raggedtensors. I just gave that a quick test though (using the unit tests you include here), and it looks like a lot of the lower-level tf.nn/other apis that Keras losses rely on don't support raggedtensors right now, so that rules out option (1) I suggested above.

I'm sure we're going to face these same challenges for other compositetensor/extensiontypes (sparsetensors, various graph-nn tensor types I know are being built out right now, etc.). So the most generic, extensible solution would be to use extensiontype dispatching to add support for these types in the `tf.nn.*` apis and/or the `keras.losses.*` methods. And it would be most consistent with the spirit of how ExtensionTypes/CompositeTensors such as RaggedTensors are intended to extend existing TF apis (https://github.com/tensorflow/community/blob/master/20200721-extension-types.md)

It would also leave us with more consistent behavior throughout the Keras + tf stack. If we just change `LossFunctionWrapper` we would be in a situation where `keras.losses.MeanSquaredError` can handle `RaggedTensors`, but `keras.losses.mean_squared_error*` apis called directly outside of fit can't, and tf.nn.* can't.

-----------
ragged_dispatch.py shows that you can do this in a generic fashion:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/ragged/ragged_dispatch.py

So, you won't have to create separate unique methods to separately handle `mse`/`mae`/`crossentropy`/etc. You should be able to register a dispatcher that does what you're generically doing in `_eval_ragged_tensors` right now, then just pass it a list of the losses you want to register RaggedTensor support for. And ideally no changes to LossFunctionWrapper itself would be needed.

Thank you again for looking into all of this & putting together what you've put together so far!

Comment (pedro-r-marques) [reply tensorflow/python/keras/losses.py:295]: The property that causes RaggedTensors to need to be treated differently is that they have different dimensions per batch; this would not be an issue that cut across all CompositeTensors. In order to add support at the tf.nn.* API level it would seem to me that it would be necessary to add a notion at the CompositeTensor base class whether the tensor has fixed dimensions (and thus can be converted to a dense tensor across batches) or has ragged dimensions and thus would need to at least be processed via ```tf.map``` on a per batch basis. If that is the direction that you would like to go in, I believe that it would require an RFC and quite a bit of consultation with tensorflow project contributors / leads; probably beyond the scope of what would be reasonable for me to attempt.

The alternative approach would be to consider this an issue to be handled at the Keras API level; which is the level that at the moment is handling the batch level logic (i.e. per batch weights and reduction policy). In that approach; I think that my suggestion of using a wrapper function that is applied via tf.map and uses the underlying loss to retrieve the per batch values is reasonable. It does add additional complexity...

it adds a logical rule that:
-  if the loss function is not already a subclass of ```losses.Loss``` and ragged tensors are being used then the library extracts a dense tensor per batch.

There are a couple of issues to consider:
- Is this an intuitive behaviour ? What would the user of the API typically expect ?
- Does it break the API for current users ?

I would argue that given that losses utils didn't support ragged tensors on a loss until #45015 that there is no significant risk of breaking existing users due to the change; the other question is more ambiguous: my personal opinion is that since the ```Loss``` class is handling the per batch aspect of the loss processing (weights, reduction), I would expect it to be able to understand ragged tensors with a ragged dimension after the batch dimension and deal with it; the metrics base class already does just that.

An additional argument is that in the presence of user defined loss functions it would still be useful for the keras losses.Loss class to provide the adaptation from ragged to dense vectors.

Please let me know your thoughts; whether you believe it is reasonable to keep pursuing a Keras loss wrapper approach or whether you believe that this is counterproductive.

I assume that another option that would not address the issue for CompositeTensors in general but could help for RaggedTensors in particular is to add the same type of wrapper I'm proposing but inside the ragged ops via dynamic dispatch;

For instance in losses.py one could add something such as:
```
@dispatch_for_types(math_ops.add, RaggedTensor, RaggedTensorValue)
def ragged_mean_squared_error(y_true, y_pred):
  ...
  result = ragged_map_ops.map_fn(mean_squared_error, elems=(y_true, y_pred), dtype=spec)
  return result
```

This would have the drawback of needing to be added once per operation... or at least one dispatch annotation per operation; it is better if different operations need different adapter functions; although imho the dynamic dispatching makes the code flow significantly harder to understand if it is the case that all operations need the same wrapper; would also not address the issue of user defined loss functions.

Comment (tomerk) [reply tensorflow/python/keras/losses.py:295]: Thanks for putting the time into this and giving it a lot of thought. Okay, I think given the LossFunctionWrapper approach:
1. is internal (the details aren't public facing)
2. doesn't rule out switching to a dispatch based approach in the future

And given our own bandwidth right now, I think it's reasonable to accept it as-is. That way the users who need it can benefit right away, and if our bandwidth opens up we can pursue a more complete dispatching approach in the future as needed. (Likely when users need support for other composite types as targets/inputs to losses).

**Review by tomerk — CHANGES_REQUESTED** [review]
Thank you again! one last request: I think these semantics should be in model.`compile`. (and perhaps noted in the base loss classes)

**Comment by pedro-r-marques** [comment]
@tomerk 
> I think these semantics should be in model.compile

Can you please clarify ?
Currently model.compile only creates the LossesContainer; from my reading of the code at this time the code doesn't know what time of tensors will be used as y_true, y_pred. The LossesContainer is only built when called with target tensors as model.fit is being used.

One potential option would be to move the RaggedTensor support from the LossFunctionWrapper class and instead into the LossesContainer; the drawback I see is that this would then no longer work for custom training loops.

The documentation of keras loss functions typically suggests creating an instance of a loss class and testing it on a tensor; for instance https://keras.io/api/losses/probabilistic_losses/#binarycrossentropy-class. From that standpoint, it seems to me that it would be preferable if the loss class itself (e.g. keras.losses.BinaryCrossentropy in our example) did support RaggedTensors when invoked.

Or did you mean that the behaviour of the loss functions needs to be documented both in the description of model.compile (which is where the user specifies the loss function that ends up getting wrapped) and in the Loss base class ? I would fully agree that it would be desirable to improve the documentation; that may alleviate slightly the concern with additional complexity / magic behaviour for RaggedTensors that we are introducing; which I do agree is a valid concern.

Apologies in advance if I'm misinterpreting your comment.

**Comment by pedro-r-marques** [comment]
@tomerk Added RaggedTensor specific info to both model.compile and losses.Loss.

**Review by tomerk — APPROVED** [review]

**Comment by pedro-r-marques** [comment]
@tomerk Previous CI failed against Google internal CI. Rebased the PR.

File: tensorflow/python/keras/losses.py (hunk 2 above)
Comment (tomerk) [reply tensorflow/python/keras/losses.py:299]: Hi @pedro-r-marques we've actually discussed this further as a team and we've decided we're not comfortable with this LossFunctionWrapper approach as a stopgap. Any handling of raggedtensors needs to be done via the loss functions themselves. The lossfunctionwrapper class should just pass compositetensors/raggedtensors directly through as is (As per our conversation above). Any custom loss will need to explicit specify how it handles raggedtensors.

We are okay with changes that:
1. explicitly update the loss function code to add raggedtensor support to the individual loss api functions (via utilities to reduce repetition)
2. use extensiontype dispatching to add raggedtensor support to the existing loss apis (as discussed above)

**Review by tomerk — CHANGES_REQUESTED** [review]
See comment about our further discussion of this as a team.

The merge test failure is unrelated though, we'll fix that on our side.

File: tensorflow/python/keras/losses.py (hunk 2 above)
Comment (pedro-r-marques) [reply tensorflow/python/keras/losses.py]: Updated the code to affect only mean_squared_error. If the approach seems reasonable then it can be extended to other losses in the library.

**Review by tomerk — APPROVED** [review]
This looks good to me, just gonna go ahead and add @omalleyt12 as well for a second opinion.

**Comment by pedro-r-marques** [comment]
@tomerk previous CI run failed due to indentation issue caught by pylint. Please re-trigger CI.

File: tensorflow/python/keras/losses.py (hunk 2 above)
Comment (omalleyt12) [reply tensorflow/python/keras/losses.py]: Thanks for the PR! This is a clever way to add `RaggedTensor` support to a loss function (and should also be the most performant in the long-run since the dispatching mechanism is moving to C++)

This LGTM, can we just change the title and description of the PR to reflect the new approach?

**Review by omalleyt12 — APPROVED** [review]
Thanks for the PR! LGTM, let's just change the PR title / description to reflect the new approach

**Comment by pedro-r-marques** [comment]
@omalleyt12 @tomerk CI instance for TF-Lite failed with error unrelated to the PR (VM install error). Please re-trigger the CI. Is there any way for me to re-trigger the CI for a PR that has been approved ? CI errors seem to be very frequent.

**Review by tomerk — APPROVED** [review]

**Review by tomerk — APPROVED** [review]
(trying to relaunch a ci run)

**Review by tomerk — CHANGES_REQUESTED** [review]
(don't need changes just trying to see if this retriggers the ci)

**Review by tomerk — APPROVED** [review]

**Review by tomerk — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Review by tomerk — APPROVED** [review]
